exec/java-exec/src/main/java/org/apache/drill/exec/ops/FragmentContext.java

```diff
@@ -17,483 +17,216 @@
+ * Allows unit testing by mocking or reimplementing services with
+ * test-time versions. The name is awkward, chosen to avoid renaming
+ * the implementation class which is used in many places in legacy code.
+ * New code should use this interface, and the names should eventually
+ * be swapped with {@link FragmentContextImpl} becoming
+ * <tt>FragmentContextImpl</tt> and this interface becoming
+ * {@link FragmentContextImpl}.
  */
 
-public class FragmentContext extends BaseFragmentContext implements AutoCloseable, UdfUtilities, FragmentContextInterface {
-  private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(FragmentContext.class);
-
-  private final Map<DrillbitEndpoint, AccountingDataTunnel> tunnels = Maps.newHashMap();
-  private final List<OperatorContextImpl> contexts = Lists.newLinkedList();
-
-  private final DrillbitContext context;
-  private final UserClientConnection connection; // is null if this context is for non-root fragment
-  private final QueryContext queryContext; // is null if this context is for non-root fragment
-  private final FragmentStats stats;
-  private final BufferAllocator allocator;
-  private final PlanFragment fragment;
-  private final ContextInformation contextInformation;
-  private IncomingBuffers buffers;
-  private final OptionManager fragmentOptions;
-  private final BufferManager bufferManager;
-  private ExecutorState executorState;
-  private final ExecutionControls executionControls;
-
-  private final SendingAccountor sendingAccountor = new SendingAccountor();
-  private final Consumer<RpcException> exceptionConsumer = new Consumer<RpcException>() {
-    @Override
-    public void accept(final RpcException e) {
-      fail(e);
-    }
-
-    @Override
-    public void interrupt(final InterruptedException e) {
-      if (shouldContinue()) {
-        logger.error("Received an unexpected interrupt while waiting for the data send to complete.", e);
-        fail(e);
-      }
-    }
-  };
-
-  private final RpcOutcomeListener<Ack> statusHandler = new StatusHandler(exceptionConsumer, sendingAccountor);
-  private final AccountingUserConnection accountingUserConnection;
-  /** Stores constants and their holders by type */
-  private final Map<String, Map<MinorType, ValueHolder>> constantValueHolderCache;
-
+public interface FragmentContext extends UdfUtilities, AutoCloseable {
   /**
-   * Create a FragmentContext instance for non-root fragment.
-   *
-   * @param dbContext DrillbitContext.
-   * @param fragment Fragment implementation.
-   * @param funcRegistry FunctionImplementationRegistry.
-   * @throws ExecutionSetupException
+   * Wait for ack that all outgoing batches have been sent
    */
-  public FragmentContext(final DrillbitContext dbContext, final PlanFragment fragment,
-      final FunctionImplementationRegistry funcRegistry) throws ExecutionSetupException {
-    this(dbContext, fragment, null, null, funcRegistry);
-  }
+  void waitForSendComplete();
```

Comment:
The goal of the interface was to define a self-contained set of services easily used in unit tests. Said another way, code that uses the interface controlled dependencies and can be unit tested without mocks.

Introducing these network-related APIs breaks that assumption. Now, any code can access network services, creating a huge API and requiring extensive mocking.

The suggestion here is to either:

1. Keep network (or server) related services only in the "impl" class (the design as of the commit that introduced this interface.)

2. Create a new "NetworkFragmentContext" that has the isolated service plus the network and server-related services.

In either case, code that uses the original interface can be easily unit tested. Code that uses the "network/server" flavor requires the full Drillbit or extensive mocking.

In general, it will only be exchange senders and receivers that need network services. All "internal" operators should be able to use the constrained version.

(Would sure be great if Drill were a project that valued design documents so all of this would not just be personal opinion or word of mouth...)

Doing the full review (see notes below), a better design might be:

1. Runtime fragment context (used by internal operators, makes it easy to unit test.)
2. Exchange fragment context (exposes network services, used only for exchange send and receive. Quite hard to unit test.)
3. Fragment builder context (used to create a fragment, but never used once the fragment is up and running.)

This three-part split will relieve the kitchen-sink design of the current classes while preserving the goal of the prior commit allow easy unit testing of internal (non-exchange) operators.

Reply:
Thanks for the explanation. Sounds very reasonable to me. I will work on splitting things up into 3 interfaces. The FragmentContextImpl class could probably implement all three interfaces and be cast to the appropriate interface depending on where it is passed to.